File: samples/python/agents/helloworld/agent_executor.py
from a2a.helpers import (
    new_task_from_user_message,
    new_text_message,
    new_text_part,
)
from a2a.server.agent_execution import AgentExecutor, RequestContext
from a2a.server.events import EventQueue
from a2a.server.tasks import TaskUpdater
from a2a.types.a2a_pb2 import TaskState


# --8<-- [start:HelloWorldAgent]
class HelloWorldAgent:
    """Hello World Agent."""

    async def invoke(self, user_request: str) -> str:
        """Invoke the Hello World agent to generate a response."""
        return f'Hello, World! I have received your request ({user_request})'


# --8<-- [end:HelloWorldAgent]


# --8<-- [start:HelloWorldAgentExecutor_init]
class HelloWorldAgentExecutor(AgentExecutor):
    """Test AgentProxy Implementation."""

    def __init__(self) -> None:
        self.agent = HelloWorldAgent()

    # --8<-- [end:HelloWorldAgentExecutor_init]

    # --8<-- [start:HelloWorldAgentExecutor_execute]
    async def execute(
        self,
        context: RequestContext,
        event_queue: EventQueue,
    ) -> None:
        """Process user request."""
        # Collect a task from request context
        if context.current_task:
            task = context.current_task
        else:
            # If there is no task, create one and add it event queue
            task = new_task_from_user_message(context.message)
            await event_queue.enqueue_event(task)

        # Update task status in EventQueue using TaskUpdater class object
        task_updater = TaskUpdater(
            event_queue=event_queue, task_id=task.id, context_id=task.context_id
        )
        await task_updater.update_status(
            state=TaskState.TASK_STATE_WORKING,
            message=new_text_message('Processing request...'),
        )

        # collect user request from content and invoke your agent to generate content
        query = ''
        for part in context.message.parts:
            if part.text:
                query += part.text
        result = await self.agent.invoke(user_request=query)

        # All generated results are added to EventQueue as artifacts
        await task_updater.add_artifact(
            parts=[new_text_part(text=result, media_type='text/plain')]
        )

        # Add the result as a task artifact and update the task status to completed
        await task_updater.update_status(
            state=TaskState.TASK_STATE_COMPLETED,
            message=new_text_message('Request is completed!'),
        )

    # --8<-- [end:HelloWorldAgentExecutor_execute]

    # --8<-- [start:HelloWorldAgentExecutor_cancel]
    async def cancel(self, context: RequestContext, event_queue: EventQueue) -> None:
        """Raise exception as cancel is not supported."""
        raise Exception('cancel not supported')

    # --8<-- [end:HelloWorldAgentExecutor_cancel]

File: samples/js/agents/helloworld/agent_executor.ts
import { v4 as uuidv4 } from 'uuid';

import {
  Artifact,
  Role,
  Task,
  TaskArtifactUpdateEvent,
  TaskState,
  TaskStatusUpdateEvent,
} from '@a2a-js/sdk';
import {
  AgentEvent,
  AgentExecutor,
  ExecutionEventBus,
  RequestContext,
  UnsupportedOperationError,
} from '@a2a-js/sdk/server';

// --8<-- [start:HelloWorldAgent]
export class HelloWorldAgent {
  /** Hello World Agent. */

  async invoke(userRequest: string): Promise<string> {
    /** Invoke the Hello World agent to generate a response. */
    return `Hello, World! I have received your request (${userRequest})`;
  }
}
// --8<-- [end:HelloWorldAgent]

// --8<-- [start:HelloWorldAgentExecutor_init]
export class HelloWorldAgentExecutor implements AgentExecutor {
  /** Test AgentProxy Implementation. */

  private readonly agent: HelloWorldAgent;

  constructor() {
    this.agent = new HelloWorldAgent();
  }
  // --8<-- [end:HelloWorldAgentExecutor_init]

  // --8<-- [start:HelloWorldAgentExecutor_execute]
  async execute(
    requestContext: RequestContext,
    eventBus: ExecutionEventBus
  ): Promise<void> {
    /** Process user request. */
    const userMessage = requestContext.userMessage;
    const existingTask = requestContext.task;
    const taskId = requestContext.taskId;
    const contextId = requestContext.contextId;

    // 1. Collect a task from request context or create snapshot
    const taskSnapshot: Task = existingTask ?? {
      id: taskId,
      contextId: contextId,
      status: {
        state: TaskState.TASK_STATE_SUBMITTED,
        timestamp: new Date().toISOString(),
        message: undefined,
      },
      artifacts: [],
      history: [userMessage],
      metadata: userMessage.metadata,
    };
    eventBus.publish(AgentEvent.task(taskSnapshot));

    // 2. Update task status to working
    const workingStatusUpdate: TaskStatusUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      status: {
        state: TaskState.TASK_STATE_WORKING,
        message: {
          role: Role.ROLE_AGENT,
          messageId: uuidv4(),
          parts: [
            {
              content: { $case: 'text', value: 'Processing request...' },
              metadata: undefined,
              filename: '',
              mediaType: 'text/plain',
            },
          ],
          taskId: taskId,
          contextId: contextId,
          extensions: [],
          metadata: {},
          referenceTaskIds: [],
        },
        timestamp: new Date().toISOString(),
      },
      metadata: {},
    };
    eventBus.publish(AgentEvent.statusUpdate(workingStatusUpdate));

    // 3. Collect user request from request content and invoke LLM agent
    const textPart = userMessage.parts.find(
      (p) => p.content?.$case === 'text'
    );
    const query =
      textPart?.content?.$case === 'text'
        ? textPart.content.value.trim()
        : '';

    let result: string;
    if (query) {
      result = await this.agent.invoke(query);
    } else {
      result = 'No text input is provided!';
    }

    // 4. Add generated response as an artifact
    const artifactId = uuidv4();
    const resultArtifact: Artifact = {
      artifactId: artifactId,
      name: 'Result',
      description: 'The response from the Hello World agent.',
      parts: [
        {
          content: { $case: 'text', value: result },
          metadata: undefined,
          filename: '',
          mediaType: 'text/plain',
        },
      ],
      metadata: undefined,
      extensions: [],
    };

    const artifactUpdate: TaskArtifactUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      artifact: resultArtifact,
      lastChunk: true,
      append: false,
      metadata: undefined,
    };
    eventBus.publish(AgentEvent.artifactUpdate(artifactUpdate));
    console.log('Result: ', result);

    // 5. Update task status to completed
    const finalUpdate: TaskStatusUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      status: {
        state: TaskState.TASK_STATE_COMPLETED,
        message: {
          role: Role.ROLE_AGENT,
          messageId: uuidv4(),
          parts: [
            {
              content: { $case: 'text', value: 'Request is completed!' },
              metadata: undefined,
              filename: '',
              mediaType: 'text/plain',
            },
          ],
          taskId: taskId,
          contextId: contextId,
          extensions: [],
          metadata: {},
          referenceTaskIds: [],
        },
        timestamp: new Date().toISOString(),
      },
      metadata: undefined,
    };
    eventBus.publish(AgentEvent.statusUpdate(finalUpdate));
  }
  // --8<-- [end:HelloWorldAgentExecutor_execute]

  // --8<-- [start:HelloWorldAgentExecutor_cancel]
  public cancelTask = async (
    _taskId: string,
    _eventBus: ExecutionEventBus
  ): Promise<void> => {
    /** Raise exception as cancel is not supported. */
    throw new UnsupportedOperationError('Cancel is not supported.');
  };
  // --8<-- [end:HelloWorldAgentExecutor_cancel]
}
